content/lectures/lec03.md

```diff
@@ -7,4 +7,192 @@ slides: ""
 draft: false
 ---
 
-Coming soon!
+# Lecture Three: Docker Compose
+
+## Limitations of Docker as a production run system
+
+We've seen how useful Docker is as a tool and the reasons that we'd want to use it in production. But it doesn't solve all of our problems. Take this `docker run` invocation, for example:
+
+```
+$ docker run -it \
+  --name=letsencrypt \
+  --cap-add=NET_ADMIN \
+  -e PUID=1000 \
+  -e PGID=1000 -e TZ=Europe/London \
+  -e URL=walthome.duckdns.org \
+  -e SUBDOMAINS=wildcard \
+  -e VALIDATION=duckdns \
+  -e DUCKDNSTOKEN=youwillnevergetthis \
+  -e EMAIL=pwpon500@gmail.com \
+  -v ./test:/config \
+  --restart unless-stopped \
+  linuxserver/letsencrypt
+```
+
+Maybe this is easy for the computer to read, but for humans there are numerous drawbacks to writing long console commands like this. For one, they're inaccessible and confusing to people who have less experience with the Docker CLI. Second, how are we going to keep track of this? Perhaps we could throw it in a shell script and commit it to our repository, but shell scripts can be cumbersome to edit and do not always interface well with other programming languages.
+
+## Run System Wishlist
+
+So, if we're unhappy with the Docker CLI as our interface for running containers, let's think about what we want out of a run system:
+
+- Reproducible (should be easy to share)
+- Easy to understand what's going on (human readable)
+- Confidence in making changes to config
+- Resource lifecycle management (we can set parameters that manage our containers)
+
+If we look at the example from the previous section, we can see that apart from resource lifecycle management, we have nothing else from our wishlist.
+
+## Docker Compose
+
+The improvement here is Docker Compose. This is a tool that allows us to "compose" multiple Docker containers together, and manage their lifecycles. Docker Compose manages the lifecycle of multiple containers at the same time, handles all runtime settings for us (environment variables, port maps, volumes, etc), and can be started, stopped and inspected with simple shell commands. Now, if we look back at our wishlist, we meet all of the criteria.
+
+## YAML (YAML Ain't Markup Language)
+
+What does Docker compose actually look like? The configuration is written in a language called YAML. This is great for a lot of reasons: YAML is a simple and human-readable config language, it is standard across many infrastructure tools like Kubernetes and Continuous Integration, and it has types which will prevent us from making some major mistakes. The grammar of YAML is made up of these base types:
+
+- Scalar
+  - String
+  - Number
+  - Boolean
+- Dictionary 
+- List
+
+The top-level structure is always a dictionary, a list of key/value pairs. If you're familiar with JSON, then this should seem familiar. While there are only a few types, by composing these data types we can be quite expressive.
```

Comment:
The top-level structure can be anything. You can have a list at the top level or just a string:
https://stackoverflow.com/questions/36663187/yaml-is-it-possible-to-have-a-list-in-the-root-section

Reply:
noted, changed to "is generally a dictionary for config files"